src/Data/List/Sufficient.agda

```diff
@@ -0,0 +1,87 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- 'Sufficient' lists: a structurally inductive view of lists xs
+-- as given by xs ≡ non-empty prefix + sufficient suffix
+--
+-- Useful for termination arguments for function definitions
+-- which provably consume a non-empty (but otherwise arbitrary) prefix
+-- *without* having to resort to ancillary WF induction on length etc.
+-- e.g. lexers, parsers etc.
+--
+-- Credited by Conor McBride as originally due to James McKinna
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.List.Sufficient {a} {A : Set a} where
+
+open import Level using (_⊔_)
+open import Data.List.Base using (List; []; _∷_; [_]; _++_)
+open import Relation.Binary.PropositionalEquality using (_≡_; refl)
+
+------------------------------------------------------------------------
+-- Sufficient builder
+
+suffAcc : ∀ {b} (B :  List A → Set b) (xs : List A) → Set (a ⊔ b)
+suffAcc B xs = ∀ {x} {prefix} suffix → xs ≡ x ∷ prefix ++ suffix → B suffix
+
+------------------------------------------------------------------------
+-- Sufficient view
+
+data Sufficient : (xs : List A) → Set a where
+
+  acc : ∀ {xs} (ih : suffAcc Sufficient xs) → Sufficient xs
+
+
+------------------------------------------------------------------------
+-- Sufficient properties
+
+-- constructors (typically not for export)
+
+module Constructors where
+
+  []⁺ : Sufficient []
+  []⁺ = acc λ _ ()
+
+  ∷⁺ : ∀ {x} {xs} → Sufficient xs → Sufficient (x ∷ xs)
+  ∷⁺ {xs = xs} suff-xs@(acc hyp) = acc λ { _ refl → suf _ refl }
```

Comment:
Can we avoid using `-` in variable names?

Reply:
Sure! I had no idea that A Thing...

Reply:
There's one more use to snag... I'll commit today if I can.

Reply:
Done.